server/src/main/java/org/elasticsearch/TransportVersions.java

```diff
@@ -225,19 +225,52 @@ static TransportVersion def(int id) {
      * In branches 8.7-8.10 see server/src/main/java/org/elasticsearch/TransportVersion.java for the equivalent definitions.
      */
 
+    static final VersionLookup VERSION_LOOKUP = ReleaseVersions.generateVersionsLookup(TransportVersions.class);
+
+    static final NavigableMap<Integer, TransportVersion> VERSION_IDS = getAllVersionIds(TransportVersions.class);
+
     /**
      * Reference to the earliest compatible transport version to this version of the codebase.
      * This should be the transport version used by the highest minor version of the previous major.
      */
-    public static final TransportVersion MINIMUM_COMPATIBLE = V_7_17_0;
+    public static final TransportVersion MINIMUM_COMPATIBLE;
 
     /**
      * Reference to the minimum transport version that can be used with CCS.
      * This should be the transport version used by the previous minor release.
      */
-    public static final TransportVersion MINIMUM_CCS_VERSION = V_8_11_X;
+    public static final TransportVersion MINIMUM_CCS_VERSION;
 
-    static final NavigableMap<Integer, TransportVersion> VERSION_IDS = getAllVersionIds(TransportVersions.class);
+    static {
+        NavigableSet<Version> allVersions = Version.getDeclaredVersions(Version.class);
+
+        Version minCompatVersion = allVersions.headSet(Version.fromString(Version.CURRENT.major + ".0.0"), false)
+            .descendingSet()
+            .stream()
+            .filter(v -> v.revision == 0)
+            .findFirst()
+            .orElseThrow(() -> new IllegalStateException("Could not find min compatible version from " + Version.CURRENT));
+
+        int tvId = VERSION_LOOKUP.findId(minCompatVersion)
+            .orElseThrow(() -> new IllegalStateException("Could not find transport version id for version " + minCompatVersion));
+        MINIMUM_COMPATIBLE = VERSION_IDS.get(tvId);
+        if (MINIMUM_COMPATIBLE == null) {
+            throw new IllegalStateException("Could not find transport version constant for id " + tvId);
+        }
+
+        if (Version.CURRENT.minor == 0) {
+            // first release of a new major - it's the same as the min compat version
+            MINIMUM_CCS_VERSION = MINIMUM_COMPATIBLE;
+        } else {
+            Version minCcsVersion = Version.fromString(Version.CURRENT.major + "." + (Version.CURRENT.minor - 1) + ".0");
+            int tvCcsId = VERSION_LOOKUP.findId(minCompatVersion)
```

Comment:
Why do we need to lookup this version? We have the Version we need above. We don't use reference equality for versions, so it won't matter if it's a different instance.

Reply:
Oops, this is using the wrong version constant. I've added a test to catch this.

Reply:
I think my comment still applies. We have minCcsVersion already, why do we convert to an id and then back to Version?

Reply:
I can't see this - we're going from the Version (maj.min-1.0) to the transport version id as stored in the csv, to the TransportVersion constant for that id. MINIMUM_CCS_VERSION is copied from MINIMUM_COMPATIBLE when its the first minor of a new major, but apart from that, we need to look up the transport version for the min compat version.

Reply:
I see my confusion, the "version" terminology overlap is breaking my brain. :)